Heterogeneous Memory Spaces

rkowalewski opened the pull request:
This pull request tracks progress about implementing heterogeneous memory spaces including the following:

- [x] Propose a Model for heterogeneous memory spaces as specified in C++17 Polymorphic Resources
- [x] Propose Memory Traits to specify capabilities on memory spaces.
- [x] Extend the model of polymorphic memory resources for global and local memory spaces
- [x] Simplify the Allocator Concept based on Allocation Policies
- [x] Provide Implementation for High Bandwidth Memory on Intel Xeon Phi KNL

- [ ] Provide CUDA Memory Space for integration with Alpakka, etc (delayed to a separate pull request)

Commit:
introducted local memory spaces

fuchsto committed:
Introducing allocator traits [no ci]

fuchsto committed:
Merge branch 'development' of github.com:dash-project/dash into feat-303-glob-dyn-alloc

fuchsto committed:
Documentation of allocator- and memory space concepts, intermediate [skip ci]

fuchsto committed:
Documentation of allocator- and memory space concepts, intermediate [skip ci]

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

Commit:
Merge remote-tracking branch 'origin/development' into feat-umap-redesign

Commit:
added Hash Policy for UnorderedMap

Commit:
added a simple memory pool for efficient allocation and deallocation

Commit:
fixed includes after moving GlobDynamicMem

Commit:
added a small unit test for SimpleMemoryPool

Commit:
Merge remote-tracking branch 'origin/feat-303-glob-dyn-alloc' into feat-umap-redesign

Conflicts:
	dash/include/dash/Memory.h
	dash/include/dash/allocator/DynamicAllocator.h
	dash/include/dash/memory/MemorySpace.h

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

Commit:
added documentation for SimpleMemoryPool

Commit:
Merge remote-tracking branch 'origin/feat-303-glob-dyn-alloc' into feat-umap-redesign

Commit:
introducing scoped local and global memory allocators

Commit:
temporarily disable unit tests

Commit:
integrate Memory Spaces Concept in Dynmic Allocator

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

Commit:
integrated allocator traits and memory spaces

Commit:
Merge remote-tracking branch 'origin/feat-303-glob-dyn-alloc' into feat-umap-redesign

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'feat-globptr' of github.com:dash-project/dash into feat-303-glob-dyn-alloc

Conflicts:
	dash/include/dash/GlobDynamicMem.h
	dash/include/dash/GlobMem.h
	dash/include/dash/Memory.h
	dash/include/dash/memory/GlobDynamicMem.h
	dash/include/dash/memory/GlobHeapMem.h
	dash/include/dash/memory/GlobMem.h
	dash/include/dash/memory/GlobStaticMem.h
	dash/include/libdash.h

fuchsto committed:
Merge branch 'feat-303-glob-dyn-alloc' of github.com:dash-project/dash into feat-303-glob-dyn-alloc

fuchsto committed:
STL-compliant definition of MemorySpace, added documentation and notes

Commit:
Make Allocators aware of the underlying memory spaces

Commit:
Merge branch 'feat-umap-redesign' into feat-303-glob-dyn-alloc

Conflicts:
	dash/include/dash/Array.h
	dash/include/dash/GlobAsyncRef.h
	dash/include/dash/GlobRef.h
	dash/include/dash/GlobSharedRef.h
	dash/include/dash/Halo.h
	dash/include/dash/List.h
	dash/include/dash/Matrix.h
	dash/include/dash/Memory.h
	dash/include/dash/Shared.h
	dash/include/dash/UnorderedMap.h
	dash/include/dash/allocator/EpochSynchronizedAllocator.h
	dash/include/dash/experimental/Halo.h
	dash/include/dash/experimental/HaloMatrix.h
	dash/include/dash/experimental/iterator/GlobStencilIter.h
	dash/include/dash/iterator/GlobStencilIter.h
	dash/include/dash/iterator/GlobViewIter.h
	dash/include/dash/list/ListRef.h
	dash/include/dash/list/LocalListRef.h
	dash/include/dash/map/UnorderedMap.h
	dash/include/dash/matrix/internal/Matrix-inl.h
	dash/include/dash/memory/MemorySpace.h
	dash/test/memory/GlobHeapMemTest.cc
	dash/test/memory/GlobStaticMemTest.cc

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

Commit:
Merge remote-tracking branch 'origin/feat-303-glob-dyn-alloc' into feat-303-glob-dyn-alloc

Commit:
Fix as reported by Clang

fuchsto committed:
Fixes for clang-5

Commit:
Merge remote-tracking branch 'origin/feat-303-glob-dyn-alloc' into feat-303-glob-dyn-alloc

fuchsto committed:
Introduced dash::max_align_t

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'development' of github.com:dash-project/dash into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'feat-303-glob-dyn-alloc' of github.com:dash-project/dash into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

Commit:
Reduce communication overhead in GlobHeapMem

Commit:
Merge branch 'feat-globptr' into feat-303-glob-dyn-globptr-alloc

Conflicts:
	dash/include/dash/GlobPtr.h
	dash/include/dash/map/UnorderedMap.h
	dash/include/dash/memory/MemorySpace.h
	dash/test/memory/GlobHeapMemTest.cc

Commit:
Minor fixes

Commit:
minor fixes in the allocators

Commit:
Merge remote-tracking branch 'origin/development' into feat-303-glob-dyn-alloc

Conflicts:
	dart-if/include/dash/dart/if/dart_globmem.h
	dash/include/dash/allocator/EpochSynchronizedAllocator.h

Commit:
fixed resize of attached memory chunks

Commit:
fixed GlobalHeapMemTest

Commit:
minor changes

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

Commit:
addresses review comments

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

fuchsto committed:
Merge branch 'development' into feat-303-glob-dyn-alloc

Commit:
Merge remote-tracking branch 'origin/development' into feat-303-glob-dyn-alloc

Commit:
Merge remote-tracking branch 'origin/development' into feat-303-glob-dyn-alloc

Commit:
Memory concept: use only local memory domain

Commit:
symmtric allocator: use dynamic windows

Commit:
Add Memkind library for High Bandwidth Memory

Commit:
Add High Bandwidth Memory Space

Commit:
Add memory space template parameter in Array to support Intel KNL HBW memory

Commit:
Unit Test for HBW Memory Space in Array

Commit:
Add memory space template parameter for Matrix

Commit:
Small Fixes

Commit:
proper alignment in min / max element with memory spaces

Commit:
remove dead code

Commit:
use High Bandwidth Memory for internal data structures as well

Commit:
fall back policy if hbw is not available

Commit:
use MPI_MODE_NOCHECK in MPI Windows

Commit:
added new collectives

Commit:
Merge remote-tracking branch 'origin/feat-dart-sort' into feat-hbw-cowichan

Commit:
Merge remote-tracking branch 'origin/feat-hbw-cowichan' into feat-hbw-cowichan

Commit:
new config for cori knl

Commit:
Merge remote-tracking branch 'dash/development' into feat-hbw-cowichan

Conflicts:
	dash/include/dash/GlobRef.h
	dash/include/dash/memory/GlobHeapMem.h
	dash/test/container/ArrayTest.cc

Commit:
[Fix] minor bug fixes, all unit tests passed

Commit:
replace posix_memalign with C11 aligned_alloc

Commit:
Merge remote-tracking branch 'dash/development' into feat-hbw-cowichan

Commit:
remove dart_alltoall functions

Commit:
register memory, even if it is null

Commit:
Merge remote-tracking branch 'origin/feat-hbw-cowichan' into feat-hbw-cowichan

Commit:
Merge remote-tracking branch 'dash/development' into feat-hbw-cowichan

Commit:
integrating memory spaces into DASH containers (does not compile)

Commit:
Merge remote-tracking branch 'origin/feat-iterator-traits' into feat-hbw-cowichan

Commit:
rework allocator and memory space concept

Commit:
fix error in symmetric allocation

Commit:
reenable ListTest

Commit:
reenable GlobStaticMemTest HBWSpace

Commit:
fix copy and move constructor

Commit:
refactoring and minor changes

Commit:
Introduce Allocator Traits

Commit:
remove unnecessary comment

Commit:
encapsulate allocation record in a separate type

Commit:
minor changes

Commit:
Reactivate UnorderedMapTest

Commit:
minor changes

Commit:
Merge remote-tracking branch 'dash/development' into feat-hbw-cowichan

Commit:
fix null pointer dereference

Commit:
enhance allocator traits

Commit:
properly construct dash::Array with default initialization

Commit:
rework of dynamic global memory

Commit:
remove dead code

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
remove dead code

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
minor changes

Commit:
fix pointer definitions in containers

Commit:
fix template parameters in HDF5

Commit:
fix explicitly defaulted constructors and assignment operators

Commit:
install missing directory as well.

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
temporary fix for the DASH side in #549

fuchsto commented:
@rkowalewski We have to coordinate our work regarding this PR, let's have a meeting please.

Commit:
refactoring to prevent code duplication

Commit:
remove unused member

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
clang-tidy fixes

Commit:
check clang and gcc versions with max_align_t issue

Commit:
gcc version < 4.9 does not define max_align_t

Commit:
fix macro check

Commit:
minor refactoring

Commit:
conceptual rework with memory spaces

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
Reimplementing GlobStaticMem

Commit:
refactoring to eliminate an inheritance layer

Commit:
ehance interface to ensure compatibility

Commit:
fix compilation error with g++ 4.8.x

Commit:
integrate refactored GlobStaticMem into DASH for testing

Commit:
revert to old school GlobStaticMem with dash::Coarray

Commit:
fix some inconsistencies with typedefs

Commit:
refactoring and renaming

Commit:
minor improvements with Memkind configuration

Commit:
delayed initialized for dash::Shared

Commit:
check for double initialization

Commit:
use default constructor

Commit:
check return value of init

Commit:
added unit test for delayed allocation

Commit:
fix missing documentation

Commit:
move includes into the cpp14 guard

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
integrate new memory spaces into dash::Shared

Commit:
separate GlobStaticMem and GlobUnitMem

Commit:
replace GlobStaticMem with NeoGlobStaticMem

Commit:
simplify memory space hierachy

Commit:
fix documentation for GlobStaticMem

Commit:
Merge remote-tracking branch 'private/feat-shared-delayed' into feat-memory-spaces

Commit:
decoupling GlobPtr and MemorySpace

Commit:
check only if DART is initialized

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
added missing include

Commit:
use dash::max_align_t instead of std::max_align_t

Commit:
minor changes

Commit:
check for self assignment in GlobRef

Commit:
remove dead code

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
memory space deal only with bytes, so no value type anymore

Commit:
added a memory space registry to map native pointers <-> segments

Commit:
remove value_type in global memory type

Commit:
ArrayTest can successfully pass; MatrixTest still not

Commit:
fix minor bug in matrix local iterator

Commit:
minor fixes with new memory space model

Commit:
use pattern member insted of original pattern reference

Commit:
check for available memory before flush

Commit:
fix for gcc 4.x

Commit:
use pattern member

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
Remove SymmetricAllocator

Commit:
add workaround for missing const correctness in Coarray

Commit:
Const correctness in converting constructor of GlobPtr

Commit:
introduce a dash::make_unique allocation routine

Commit:
remove RawDartPointer as it is useless now

Commit:
remove GlobConstPtr

Commit:
remove dash::distance of two raw dart pointers

Commit:
hopefully a fix for gcc 4.x in CI

Commit:
Store GlobStaticMem instances into the MemorySpaceRegistry

Commit:
make dash::distance more efficient, eliminate recursion

Commit:
refactoring to improve readability

Commit:
move iteration logic out of GlobPtr

Commit:
implement increment for pointers in noncontig memory

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
fix const correctness in dash::copy_async

Commit:
fix log error after resolving merge conflict

Commit:
minor fixes in local memory spaces

Commit:
use unique_ptr to manage global memory resource in Array

Commit:
rename make_unique -> allocate_unique

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
remove c++14 make_unique implementation

Commit:
rename cpp17 namespace to std namespace

Commit:
added is_nothrow_swappable trait (C++17)

Commit:
fixing move semantics with memory registry. enhance tests

Commit:
remove dead code

Commit:
try if removal of c++11 enforcement solves the problem

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
RAII for dash::Shared

Commit:
Adapt memory allocation in matrix to new concept

Commit:
remove native pointer to memory space from GlobPtr

Commit:
simplify typedefs

Commit:
remove dead code

Commit:
add dash::addressof to obtain a GlobPtr from a GlobRef

Commit:
minor changes in global memory tests

devreal commented on CMakeLists.txt:
```diff
@@ -376,6 +381,15 @@ if (ENABLE_HDF5)
 endif()
+if (ENABLE_MEMKIND)
+  if (MEMKIND_FOUND)
+    message(INFO "MEMKIND                   enabled")
```
Should be `message(NOTE ...`

rkowalewski replied:
hm...the other's have `INFO` as well if enabled.

devreal replied:
Ahh you're right

devreal commented on dash/examples/app.01.mindeg/main.cpp:
```diff
@@ -13,6 +13,8 @@ using std::cout;
 using std::cin;
 using std::endl;
 
+using memalloc_ptr_t = decltype(dash::memalloc<int>(size_t{}));
```
This seems to be pattern now. Can we provide a wrapper to avoid the use of `decltype`?

rkowalewski replied:
We use `dash::memalloc` not too often. But yes I can give it a try. Even though I think it's not a bad pattern.

devreal replied:
Something along the lines of `dash::memalloc_ptr_t<T>` should be easy to do

rkowalewski replied:
You convinced me, it is more readable. Done.

devreal commented on dash/include/dash/Array.h:
```diff
@@ -423,14 +420,15 @@ class AsyncArrayRef
 
  private:
-  Array<T, index_type, PatternT>                  * _arr;
+  Array<T, index_type, PatternT, MSpaceC>                  * _arr;
```
Fix indentation

devreal commented on dash/include/dash/Array.h:
```diff
@@ -944,35 +943,44 @@ class Array
+    // b) swap memory resources
+    std::swap(m_globmem, other.m_globmem);
+    // c) move the unique_gptr from other into this
+    unique_gptr_t __tmp{other.m_data.release(),
```
This also seems to be a pattern: can we define a `swap` that encapsulates this?

rkowalewski replied:
The only way to do that is to implemented my own `dash::unique_ptr`. And that is worse compared to this IMHO. `std::unique_ptr` is limited somehow with it's API and this approach here is also the recommended from the `std::unique_ptr` [author](https://stackoverflow.com/questions/9167205/why-doesnt-unique-ptrreset-have-overloads-that-take-a-deleter/9167455#9167455).

devreal commented on dash/include/dash/GlobPtr.h:
```diff
@@ -154,93 +98,83 @@ class GlobPtr
    */
   template <typename T, class MemSpaceT>
-  self_t & operator=(const GlobPtr<T, MemSpaceT> & other)
+  self_t &operator=(const GlobPtr<T, MemSpaceT> &other)
```
Shouldnt't we check for const correctness and assignability here, too?

rkowalewski replied:
This copy constructor is now deleted anyway. I will add it in future by considering the correct memory traits.

devreal commented on dash/include/dash/GlobPtr.h:
```diff
@@ -319,11 +246,12 @@ class GlobPtr
       increment(n);
-    } else {
+    }
+    else {
```
This change seems purely cosmetic (and frankly less attractive imho)

devreal commented on dash/include/dash/GlobRef.h:
```diff
@@ -441,10 +421,19 @@ std::ostream & operator<<(
 }
 
+/**
+ * specialization for unqualified calls to swap
```
Fix documentation

devreal commented on dash/include/dash/Shared.h:
```diff
@@ -182,9 +214,9 @@ class Shared {
-    DASH_ASSERT(!DART_GPTR_ISNULL(m_dart_gptr));
-    return reference(m_dart_gptr);
+    DASH_LOG_DEBUG_VAR("Shared.get", m_glob_pointer);
+    DASH_ASSERT(static_cast<bool>(m_glob_pointer));
```
Is the static cast really needed? (It's missing in line 230)

rkowalewski replied:
Removed.

devreal commented on dash/include/dash/Types.h:
```diff
@@ -344,6 +347,21 @@ constexpr team_unit_t   UNDEFINED_TEAM_UNIT_ID{DART_UNDEFINED_UNIT_ID};
+      __attribute__((__aligned__(__alignof__(long long))));
+  long double __max_align_ld
+      __attribute__((__aligned__(__alignof__(long double))));
+    }
```
Is this still needed since we moved to C++14? Also, the struct will be larger than 16 Byte (32B on 64bit systems, 8 for `__max_align_ll` + 16 for `__max_align_ld`)

rkowalewski replied:
Good points. Let's remove it.

devreal commented on dash/include/dash/algorithm/Find.h:
```diff
@@ -50,17 +50,18 @@ GlobIter find(
-    if(l_result == l_range_end){
-      DASH_LOG_DEBUG("Not found in local range");
-      g_index = std::numeric_limits<p_index_t>::max();
+   const auto* lbegin = static_cast<typename GlobIter::const_local_type>(
```
Is that `static_cast` really needed?

rkowalewski replied:
Changed now.

devreal commented on dash/include/dash/algorithm/MinMax.h:
```diff
@@ -314,17 +327,16 @@ GlobInputIt min_element(
   /// Element comparison function, defaults to std::less
   Compare                                    compare
-    = std::greater<const ElementType &>())
+    = Compare())
```
The formatting seems weird: put both on one line for better readability

rkowalewski replied:
Yup, good point. I used `sed` sometimes, maybe that is the reason.

devreal commented on dash/include/dash/algorithm/MinMax.h:
```diff
@@ -357,7 +369,7 @@ const ElementType * max_element(
   /// Element comparison function, defaults to std::less
   Compare             compare
-    = std::greater<ElementType &>())
+    = Compare())
```
See above

devreal commented on dash/include/dash/allocator/AllocationPolicy.h:
```diff
@@ -0,0 +1,422 @@
+
+enum class global_allocation_policy : uint8_t {
+  /// All units allocate the same amount of memory
+  collective,
```
There are cases where units allocate collectively but no the same amount of memory. Should that be a separate allocation policy?

fmoessbauer replied:
@devreal: In my oppinion not as it is a different category, but do we test that?

devreal replied:
At the moment we do not test that but that is something I've been meaning to add for a while. MPI privides an info key that can steer some optimizations inside the MPI implementation if all units are guaranteed to allocate the same amount of memory but I'm not sure how big the influence is.

rkowalewski replied:
> There are cases where units allocate collectively but no the same amount of memory. Should that be a separate allocation policy?

We test that for example with the CSR Pattern. For example in the `SortTest`, we show a use case where some units take part in the collective global allocation but contribute 0 "local" bytes.

rkowalewski replied:
collective allocation is just a matter of synchronization. It does not say anything about the capacity. That's what MPI provides us.

devreal replied:
I see, I guess I am puzzled by the documentation `All units allocate the same amount of memory`. I guess what you mean is `All units collectively allocate memory`?

rkowalewski replied:
Ah correct. Now I understand your question.

rkowalewski replied:
Documentation updated.

devreal commented on dash/include/dash/allocator/AllocationPolicy.h:
```diff
@@ -0,0 +1,422 @@
+ * release the memory while other units are still operating on it.
+ */
+template <class LocalAlloc, class LMemSpace>
+class LocalAllocationPolicy {
```
Out of curiosity: how would we fit the DART allocators in here? Would that be passed through the `LocalAlloc` argument passed to `do_global_allocate`?

devreal commented on dash/include/dash/allocator/internal/Types.h:
```diff
@@ -0,0 +1,133 @@
+namespace allocator {
+
+template <typename LPtr, typename GPtr = dart_gptr_t>
+struct allocation_rec {
```
What is an `allocation_rec`? Allocation record? Please document

rkowalewski replied:
You are right. It is actually only an internal type (internal namespace) and will be removed soon. It is used only in the EpochSynchronizedAllocator. And this will be changed to follow the new memory space concept in another pull request (near future). At the moment, only static containers are adapted to it. And this was a significant amount of work.

devreal commented on dash/include/dash/iterator/internal/GlobPtrBase.h:
```diff
@@ -0,0 +1,366 @@
+    MemorySpaceT const *mem_space,
+    memory_space_noncontiguous)
+{
+  DASH_THROW(dash::exception::NotImplemented, "not implemented yet");
```
Use `static_assert` here to avoid runtime surprises?

rkowalewski replied:
In this case, not required actually. :)

devreal commented on dash/src/memory/HostSpace.cc:
```diff
@@ -0,0 +1,15 @@
+#include <dash/memory/HostSpace.h>
```
Out of curiosity: why are the memory spaces not implemented in the header? Is there some significant compilation overhead?

rkowalewski replied:
Because it does not rely on templates. ;)

devreal commented on dash/src/memory/internal/MemorySpaceRegistry.cc:
```diff
@@ -0,0 +1,72 @@
+  m_segments.erase(it);
+}
+
+std::vector<dash::internal::MemorySpaceRegistry::element_t>::iterator
```
Remove `dash::internal` from `dash::internal::MemorySpaceRegistry::element_t`

rkowalewski replied:
Ok should be done now.

devreal commented on dash/include/dash/memory/internal/MemorySpaceRegistry.h:
```diff
@@ -0,0 +1,60 @@
+
+private:
+  static std::unique_ptr<MemorySpaceRegistry> m_instance;
+  static std::vector<element_t>               m_segments;
```
Why not use a ` std::unordered_map` and hash `segid_t` and `teamid_t` into a 32-bit integer? The linear search may otherwise become significant for larger numbers of segments, e.g., when using the dynamic data structures.

rkowalewski replied:
std::vector is way faster for a small number of entries than `unordered_map`. We can change it in future but for now I would stay with a `std::vector` because we have rarely more than 10 allocated global memory instances in a single DASH application.

devreal reviewed (changes requested):
Thanks a lot for all the changes, definitely many good improvements in here! Some comments inline.

Commit:
conceptual minor changes to improve memory spaces and allocators

Commit:
review comments

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
formatting changes

Commit:
cleanup in local allocation policy

Commit:
fix documentation, review comments

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
added a type alias for convenience - review comment

Commit:
added a unit test for array of global pointers

Commit:
fix test for > 2 processes

rkowalewski commented:
This pull request contains now a significantly refined implementation of memory spaces. We have now a clear separation of concerns between memory, allocators, containers and pointers. I will provide a more enhanced documentation in our internal wiki. Just to summarize:

- Passing around pointers in global memory: Pointer arithmetic, deferencing, etc. works correct on any global pointer on any unit, as defined by the memory traits.
- Memory has no `value_type` anymore as memory knows only about bytes (i.e. `std::byte`)
- Allocators use memory to allocate / deallocate objects according to the alignment requirements.
- Containers use now always RAII by wrapping the allocated memory segment in a `std::unique_ptr`.
- I removed lots of code, fixed pointer conversion and, `dash::distance` and pointer arithmetics in general.

rkowalewski commented:
As presented in the Stuttgart meeting, we can now allocate in any memory type. This pull request supports RAM and HBM (on Intel KNL).

CUDA and non-volatile (i.e. persistent memory) follow in a separate pull request. Just to limit the scope for this pull request here.

Commit:
Merge branch 'sup-compiler-warnings' into feat-memory-spaces

Commit:
fix narrowing conversion

fmoessbauer commented on dash/examples/ex.04.memalloc/main.cpp:
```diff
@@ -19,20 +19,22 @@ int main(int argc, char* argv[])
   auto size = dash::size();
 
-  dash::Array< dash::GlobPtr<int> > arr(size);
+  using pointer_t = decltype(dash::memalloc<int>(size_t{}));
```
Can that be replaced with `dash::GlobMemAllocPtr<int>`? In an example this looks cleaner.

rkowalewski replied:
Must have overseen that. Done.

fmoessbauer commented on dash/include/dash/Matrix.h:
```diff
@@ -95,23 +96,31 @@ template <
+  typename LocalMemSpaceT>
 class LocalMatrixRef;
 
+namespace halo {
```
Why is this added to the `Matrix` class?

rkowalewski replied:
Because we can then forward declare `HaloMatrixWrapper` and make it a friend. Otherwise c++ does not allow that.

fmoessbauer commented on dash/include/dash/coarray/CoEventIter.h:
```diff
@@ -13,7 +13,12 @@ namespace coarray {
-  using gptr_t = GlobPtr<dash::Atomic<int>>;
+  //TODO rko: fix this hard coded information it is related to Coevent
+  //which uses dash::Array internally
+  using globmem_t =
```
Can't we use a `decltype` here?

fmoessbauer commented on dash/include/dash/coarray/CoEventRef.h:
```diff
@@ -5,14 +5,21 @@
-  using gptr_t      = GlobPtr<event_ctr_t>;
+  //TODO rko: fix this hard coded information it is related to Coevent
+  //which uses dash::Array internally
+  using globmem_t =
```
Same here

rkowalewski replied:
Hmm...not sure what you mean? The only way I see is to provide the pointer type as a template parameter to `CoEventIter`.

fmoessbauer commented on dash/include/dash/internal/Math.h:
```diff
@@ -3,52 +3,85 @@
 namespace math {
 
+template <unsigned Modulus, typename Integer>
+constexpr Integer RoundDown(Integer i)
```
Missing documentation, strange name.

rkowalewski replied:
Removed as it is not used anymore. ;)

fmoessbauer reviewed (commented):
There are many good things in this PR. However the amount of changes is massive. I'm not sure if a full review is actually possible. See my inline comments. After @devreal has approved, we can merge IMO.

Commit:
review comments

rkowalewski commented:
Thank you for all your review comments.

Commit:
Merge branch 'sup-compiler-warnings' into feat-memory-spaces-compiler-warnings

Commit:
Merge branch 'feat-memory-spaces' into feat-memory-spaces-compiler-warnings

Commit:
Merge branch 'sup-compiler-warnings' into feat-memory-spaces-compiler-warnings

Commit:
clarify documentation [skip-ci]

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces

Commit:
delete unused file

Commit:
Merge remote-tracking branch 'dash/development' into feat-memory-spaces-compiler-warnings

Commit:
Merge branch 'feat-memory-spaces-compiler-warnings' into feat-memory-spaces

Commit:
make Future noexcept

Commit:
use correct traits to determine noexcept guarantees

Commit:
fix move assignment and move constructor

devreal commented:
@rkowalewski Thanks for the great work!
